File: practice/src/capture_and_store.cpp
/*************************************************
This program captures frames from camera and stores
them in a file after downsampling the video frames
A trackbar is included which decides the
downsampling from 2 to 8.

author: Anshay
Date: 23/12/2020
*************************************************/

#include <iostream>
#include <opencv4/opencv2/opencv.hpp>

int g_max_downsampling = 2;
int g_slider_pos = 0;

int main(int argc, char** argv){
    cv::VideoCapture cap;
    cv::VideoWriter writer;
    std::string WINDOW_NAME = "Video Camera";
    std::string WINDOW_NAME2 = "Downsampled";
    cv::namedWindow(WINDOW_NAME);
    cv::namedWindow(WINDOW_NAME2);
    cv::Mat frame;
    // error handling
    if(argc!=2){
        std::cout<<"\nUsage capture_and_store <output_file_path>\n";
        return -1;
    }
    cap.open(0);
    if (!cap.isOpened()) {
        std::cerr << "ERROR! Unable to open camera\n";
        return -1;
    }
    cap >> frame;
    if (frame.empty()) {
        std::cerr << "ERROR! blank frame grabbed\n";
        return -1;
    }
    // VideoWriter
    cv::Size size((int)cv::CAP_PROP_FRAME_WIDTH/2, (int)cv::CAP_PROP_FRAME_HEIGHT/2);
    double fps = 25.0;
    writer.open(argv[1], cv::VideoWriter::fourcc('M','J','P','G'), fps, size);
    if (!writer.isOpened()) {
        std::cerr << "Could not open the output video file for write\n";
        return -1;
    }
    // create a trackbar
    cv::createTrackbar("Downsampling", WINDOW_NAME, &g_slider_pos,
                        g_max_downsampling);
    // loop the frames
    while(true){
        cap >> frame;
        if (frame.empty()) {
            std::cerr << "ERROR! blank frame grabbed\n";
            return -1;
        }
        cv::imshow(WINDOW_NAME, frame);
        // downsample the frame
        cv::pyrDown(frame, frame);
        writer << frame;
        for(int i=0; i<g_slider_pos; i++){
            cv::pyrDown(frame, frame);
        }
        cv::imshow(WINDOW_NAME2, frame);
        if(cv::waitKey(5)>=0) break;
    }
    // destroy windows
    cv::destroyWindow(WINDOW_NAME);
    cv::destroyWindow(WINDOW_NAME2);
    return 0;
}

File: practice/scripts/capture_and_store.py
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


# This program captures frames from camera and stores
# them in a file after downsampling the video frames
# A trackbar is included which decides the
# downsampling from 2 to 8.
# 
# author: Anshay
# Date: 06/01/2021


# In[ ]:


import numpy as np
import cv2
import sys


# In[ ]:


def scaleFrame(frame, trackbar_name, window_name):
    factor = cv2.getTrackbarPos(trackbar_name, window_name)
    while(factor>0):
        frame = cv2.pyrDown(frame)
        factor -= 1
    return frame


# In[ ]:


def capture_and_store(FILENAME):
    window_name = "Capture"
    trackbar_name = "Downsampling"
    cv2.namedWindow(window_name, cv2.WINDOW_NORMAL)
    cap = cv2.VideoCapture()
    writer = cv2.VideoWriter()
    cap.open(0)
    if(not cap.isOpened()):
        print("\nCould not open camera")
        exit()
    try:
        cv2.createTrackbar(trackbar_name, window_name, 0, 3, (lambda a: None))
        ret, frame = cap.read()
        fps = 24.0
        size = (frame.shape[1],frame.shape[0])
        writer.open(FILENAME, cv2.VideoWriter_fourcc(*"MJPG"), fps, size)
        while(ret):
            frameScaled = scaleFrame(frame, trackbar_name, window_name)
            cv2.imshow(window_name, frameScaled)
            writer.write(frame)
            if(cv2.waitKey(33)==27):
                print("\nExiting and closing camera")
                break
            ret, frame = cap.read()
    finally:
        writer.release()
        cap.release()


# In[ ]:


if __name__ == "__main__":
    if(len(sys.argv) != 2):
        print("\nCorrect usage: capture_and_store <output_file_path>\n")
        exit()
    filename = sys.argv[1]
    capture_and_store(filename)
    cv2.destroyAllWindows()


# In[ ]:





# In[ ]:




